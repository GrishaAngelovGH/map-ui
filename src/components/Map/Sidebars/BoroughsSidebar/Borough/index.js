import Borough from './Borough'

export default Borough