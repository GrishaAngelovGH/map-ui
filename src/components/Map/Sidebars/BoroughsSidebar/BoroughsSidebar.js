import { useContext } from 'react'
import LondonAreaContext from '../../../../contexts/LondonAreaContext'
import Sidebar from '../../../Sidebar'

const BoroughsSidebar = ({ showSidebar, onHide }) => {
  const londonArea = useContext(LondonAreaContext)

  const body = (
    <div>
      {
        londonArea.features.map(v => {
          const id = v.properties['cartodb_id']

          return (
            <div key={id}>
              <h5>#{id.toString().padStart(2, '0')} {v.properties.name}</h5>
            </div>
          )
        })
      }
    </div>
  )

  return (
    <Sidebar
      title='London Boroughs'
      body={body}
      showSidebar={showSidebar}
      onHide={onHide}
    />
  )
}

export default BoroughsSidebar