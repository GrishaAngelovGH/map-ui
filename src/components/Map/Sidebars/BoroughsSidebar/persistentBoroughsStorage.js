const persistentBoroughsStorage = {
  init: () => {
    const favoriteBoroughs = window.localStorage.getItem('boroughs')
    if (!favoriteBoroughs) window.localStorage.setItem('boroughs', '')
  },
  clearBoroughs: () => {
    window.localStorage.setItem('boroughs', '')
  },
  getFavoritesCount: () => {
    const favoriteBoroughs = window.localStorage.getItem('boroughs')
    return favoriteBoroughs.split(',').length - 1
  },
  isFavorite: name => {
    return window.localStorage.getItem('boroughs').includes(name)
  },
  addToFavorites: name => {
    const favoriteBoroughs = window.localStorage.getItem('boroughs')

    if (!favoriteBoroughs.includes(name)) {
      window.localStorage.setItem('boroughs', `${favoriteBoroughs}${name},`)
    }
  },
  removeFromFavorites: name => {
    const favoriteBoroughs = window.localStorage.getItem('boroughs').split(',')
    const newFavoriteBoroughs = favoriteBoroughs.filter(v => v !== name)
    window.localStorage.setItem('boroughs', newFavoriteBoroughs)
  }
}

export default persistentBoroughsStorage