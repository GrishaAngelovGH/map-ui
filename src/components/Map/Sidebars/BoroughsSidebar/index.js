import BoroughsSidebar from './BoroughsSidebar'

export default BoroughsSidebar