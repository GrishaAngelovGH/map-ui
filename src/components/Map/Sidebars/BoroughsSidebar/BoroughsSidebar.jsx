import { useContext, useEffect, useState } from 'react'

import RadioButtonGroup from 'components/RadioButtonGroup'
import SearchInput from 'components/SearchInput'
import Sidebar from 'components/Sidebar'
import Borough from './Borough'

import LondonAreaContext from 'contexts/LondonAreaContext'
import useSearch from 'hooks/useSearch'

import persistentBoroughsStorage from './persistentBoroughsStorage'

const BoroughsSidebar = ({ showSidebar, onHide, onBoroughClick }) => {
  const [viewCriteria, setViewCriteria] = useState('all')
  const [sortCriteria, setSortCriteria] = useState('id')
  const [searchCriteria, setSearchCriteria] = useState('')
  const [boroughId, setBoroughId] = useState(0)
  const [favoritesCount, setFavoritesCount] = useState(0)
  const londonArea = useContext(LondonAreaContext)

  const allBoroughs = viewCriteria === 'all' ?
    londonArea.features :
    londonArea.features.filter(v => persistentBoroughsStorage.isFavorite(v.properties.name))

  const [boroughs, boroughsCount] = useSearch(searchCriteria, allBoroughs, 'properties.name')

  const handleViewOnMapClick = (id, coords) => {
    setBoroughId(id)
    onBoroughClick(coords)
  }

  const handleClearFavoritesClick = () => {
    persistentBoroughsStorage.clearBoroughs()
    countFavorites()
  }

  const countFavorites = () => {
    setFavoritesCount(persistentBoroughsStorage.getFavoritesCount())
  }

  useEffect(() => {
    persistentBoroughsStorage.init()
    countFavorites()
  }, [])

  const radioButtonGroups = [
    {
      label: 'View',
      radioButtonGroup: (
        <RadioButtonGroup
          initialValue='all'
          radios={[
            { name: 'All', value: 'all' },
            { name: `Favorites (${favoritesCount})`, value: 'favorites' }
          ]}
          onClick={setViewCriteria}
        />
      )
    },
    {
      label: 'Sort ASC By',
      radioButtonGroup: (
        <RadioButtonGroup
          initialValue='id'
          radios={[
            { name: 'Id', value: 'id' },
            { name: 'Name', value: 'name' },
            { name: 'Area', value: 'area_hectares' }
          ]}
          onClick={setSortCriteria}
        />
      )
    }
  ]

  const body = (
    <div className='row p-1'>
      <div className='col-md-12'>
        <SearchInput placeholder='Search by name' onChange={setSearchCriteria} />

        {
          radioButtonGroups.map((v, i) => (
            <div key={i} className='row mb-3 justify-content-center align-items-center'>
              <div className='col-md-4'>
                <span>{v.label}</span>
              </div>
              <div className='col-md-7'>
                {v.radioButtonGroup}
              </div>
            </div>
          ))
        }

        {
          favoritesCount > 0 && (
            <button
              className='btn btn-danger w-100 mb-3 d-flex justify-content-center align-items-center'
              onClick={handleClearFavoritesClick}
            >
              <span className='mx-2'>Clear Favorites</span>
              <i className='bi bi-trash3 fs-4'></i>
            </button>
          )
        }

        {
          boroughs
            .sort((a, b) => {
              if (a.properties[sortCriteria] > b.properties[sortCriteria]) return 1
              if (a.properties[sortCriteria] < b.properties[sortCriteria]) return -1
              return 0
            })
            .map(v => {
              const id = v.properties.id

              return (
                <Borough
                  key={id}
                  id={id}
                  active={id === boroughId}
                  name={v.properties.name}
                  code={v.properties.code}
                  areaHectares={v.properties.area_hectares}
                  coordinates={v.geometry.coordinates[0][0]}
                  isFavorite={persistentBoroughsStorage.isFavorite(v.properties.name)}
                  onViewOnMapClick={handleViewOnMapClick}
                  onFavoriteChange={countFavorites}
                />
              )
            })
        }
      </div>
    </div>
  )

  return (
    <Sidebar
      title={<h3>London Boroughs ({boroughsCount})</h3>}
      body={body}
      showSidebar={showSidebar}
      onHide={onHide}
    />
  )
}

export default BoroughsSidebar