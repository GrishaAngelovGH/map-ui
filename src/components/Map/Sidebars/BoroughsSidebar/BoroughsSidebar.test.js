import { render } from '@testing-library/react'

import BoroughsSidebar from './BoroughsSidebar'

test('should render BoroughsSidebar component', () => {
  const view = render(<BoroughsSidebar showSidebar />)

  expect(view).toMatchSnapshot()
})