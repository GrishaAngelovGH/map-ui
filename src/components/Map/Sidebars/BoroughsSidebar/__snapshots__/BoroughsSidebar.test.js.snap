// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BoroughsSidebar component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="modal-open"
    data-rr-ui-modal-open=""
    style="padding-right: 1024px;"
  >
    <div />
    <div
      aria-modal="true"
      class="offcanvas offcanvas-start offcanvas-toggling show"
      role="dialog"
      style="visibility: visible;"
      tabindex="-1"
    >
      <div
        class="offcanvas-header"
      >
        <div
          class="offcanvas-title h5"
        >
          London Boroughs
        </div>
        <button
          aria-label="Close"
          class="btn-close"
          type="button"
        />
      </div>
      <div
        class="offcanvas-body"
      >
        <div>
          <div>
            <h5>
              #
              01
               
              Barking and Dagenham
            </h5>
          </div>
          <div>
            <h5>
              #
              02
               
              Barnet
            </h5>
          </div>
          <div>
            <h5>
              #
              03
               
              Bexley
            </h5>
          </div>
          <div>
            <h5>
              #
              04
               
              Brent
            </h5>
          </div>
          <div>
            <h5>
              #
              05
               
              Bromley
            </h5>
          </div>
          <div>
            <h5>
              #
              06
               
              Camden
            </h5>
          </div>
          <div>
            <h5>
              #
              07
               
              City of London
            </h5>
          </div>
          <div>
            <h5>
              #
              08
               
              Ealing
            </h5>
          </div>
          <div>
            <h5>
              #
              09
               
              Enfield
            </h5>
          </div>
          <div>
            <h5>
              #
              10
               
              Greenwich
            </h5>
          </div>
          <div>
            <h5>
              #
              11
               
              Hackney
            </h5>
          </div>
          <div>
            <h5>
              #
              12
               
              Hammersmith and Fulham
            </h5>
          </div>
          <div>
            <h5>
              #
              13
               
              Haringey
            </h5>
          </div>
          <div>
            <h5>
              #
              14
               
              Harrow
            </h5>
          </div>
          <div>
            <h5>
              #
              15
               
              Havering
            </h5>
          </div>
          <div>
            <h5>
              #
              16
               
              Hillingdon
            </h5>
          </div>
          <div>
            <h5>
              #
              17
               
              Hounslow
            </h5>
          </div>
          <div>
            <h5>
              #
              18
               
              Islington
            </h5>
          </div>
          <div>
            <h5>
              #
              19
               
              Kensington and Chelsea
            </h5>
          </div>
          <div>
            <h5>
              #
              20
               
              Kingston upon Thames
            </h5>
          </div>
          <div>
            <h5>
              #
              21
               
              Lambeth
            </h5>
          </div>
          <div>
            <h5>
              #
              22
               
              Lewisham
            </h5>
          </div>
          <div>
            <h5>
              #
              23
               
              Merton
            </h5>
          </div>
          <div>
            <h5>
              #
              24
               
              Newham
            </h5>
          </div>
          <div>
            <h5>
              #
              25
               
              Redbridge
            </h5>
          </div>
          <div>
            <h5>
              #
              26
               
              Richmond upon Thames
            </h5>
          </div>
          <div>
            <h5>
              #
              27
               
              Southwark
            </h5>
          </div>
          <div>
            <h5>
              #
              28
               
              Sutton
            </h5>
          </div>
          <div>
            <h5>
              #
              29
               
              Tower Hamlets
            </h5>
          </div>
          <div>
            <h5>
              #
              30
               
              Waltham Forest
            </h5>
          </div>
          <div>
            <h5>
              #
              31
               
              Wandsworth
            </h5>
          </div>
          <div>
            <h5>
              #
              32
               
              Westminster
            </h5>
          </div>
          <div>
            <h5>
              #
              33
               
              Croydon
            </h5>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
